config/tools/man/ccs_tool.8
.TH "ccs_tool" "8" "" "" ""
.SH "NAME"
ccs_tool \- The tool used to make online updates of CCS config files.

.SH "SYNOPSIS"
.B ccs_tool
[\fIOPTION\fR].. <\fBcommand\fP>

.SH "DESCRIPTION"

\fBccs_tool\fP is part of the Cluster Configuration System (CCS).  It is
used to make online updates to cluster.conf.  It can also be used to
upgrade old style (GFS <= 6.0) CCS archives to the new xml cluster.conf
format.

.SH "OPTIONS"
.TP 
\fB\-h\fP
Help.  Print out the usage.
.TP 
\fB\-V\fP
Print the version information.

sub\-commands have their own options, see below for more detail
.SH "COMMANDS"

.TP 
\fBaddnode\fP [options] \fI<node> [<fenceoption=value>]...\fP
Adds a new node to the cluster configuration file. Fencing device options
are specified as key=value pairs (as many as required) and are entered into the
configuration file as is. See the documentation for your fencing agent for more
details (eg a powerswitch fence device may need to know which port the node is
connected to).
.br
\fIOptions:\fP
.br
\-v <votes>        Number of votes for this node (mandatory)
.br 
\-n <nodeid>       Node id for this node (optional)
.br
\-i <interface>    Network interface to use for this node. Mandatory if the cluster 
is using multicast as transport. Forbidden if not.
.br
\-m <multicast>    Multicast address for cluster. Only allowed on the first node to 
be added to the file. Subsequent nodes will use either multicast or broadcast 
depending on the properties of the first node.
.br
\-f <fencedevice>  Name of fence device to use for this node. The fence device 
section must already have been added to the file, probably using the addfence command.
.br
\-c <file>         Config file to use. Defaults to /etc/cluster/cluster.conf
.br
\-o <file>         Output file. Defaults to the same as -c
.br
\-C                Don't run "ccs_tool update" after changing file. This will
happen by default if the input file is the same as the output file.
.br
\-F                Force a "ccs_tool update" even if the input and output files
are different.



.TP 
\fBdelnode\fP [options] \fI<node>\fP
Delete a node from the cluster configuration file. Note: there is no 
"edit" command so to change the properties of a node you must delete it
and add it back in with the new properties.
.br
\fIOptions:\fP
.br
\-c <file>         Config file to use. Defaults to /etc/cluster/cluster.conf
.br
\-o <file>         Output file. Defaults to the same as -c
.br
\-C                Don't run "ccs_tool update" after changing file. This will
happen by default if the input file is the same as the output file.
.br
\-F                Force a "ccs_tool update" even if the input and output files
are different.



.TP 
\fBaddfence\fP [options] \fI<name> <agent> [<option>=<value>]...\fP
Adds a new fence device section to the cluster configuration file. <agent> is the
name of the fence agent that controls the device. the options following are entered
as key-value pairs. See the fence agent documentation for details about these. eg:
you may need to enter the IP address and username/password for a powerswitch fencing
device.
.br
\fIOptions:\fP
.br
\-c <file>         Config file to use. Defaults to /etc/cluster/cluster.conf
.br
\-o <file>         Output file. Defaults to the same as -c
.br
\-C                Don't run "ccs_tool update" after changing file. This will
happen by default if the input file is the same as the output file.
.br
\-F                Force a "ccs_tool update" even if the input and output files
are different.

.TP 
\fBdelfence\fP [options] \fI<node>\fP
Deletes a fencing device from the cluster configuration file.
delfence will allow you to remove a fence device that is in use by nodes.
This is to allow changes to be made, but be aware that it may produce an
invalid configuration file if you don't add it back in again.
.br
\fIOptions:\fP
.br
\-c <file>         Config file to use. Defaults to /etc/cluster/cluster.conf
.br
\-o <file>         Output file. Defaults to the same as -c
.br
\-C                Don't run "ccs_tool update" after changing file. This will
happen by default if the input file is the same as the output file.
.br
\-F                Force a "ccs_tool update" even if the input and output files
are different.


.TP 
\fBlsnode [options] \fP
List the nodes in the configuration file. This is (hopefully obviously) not
necessarily the same as the nodes currently in the cluster, but it should
be a superset.
.br
\fIOptions:\fP
.br
\-v                Verbose. Lists all the properties of the node, and the
node-specific properties of the fence device too.
.br
\-c <file>         Config file to use. Defaults to /etc/cluster/cluster.conf


.TP 
\fBlsfence [options] \fP
List all the fence devices in the cluster configuration file.
.br
\fIOptions:\fP
.br
\-v                Verbose. Lists all the properties of the fence device rather
than just the names and agents.
.br
\-c <file>         Config file to use. Defaults to /etc/cluster/cluster.conf


.TP 
\fBcreate [options] \fP \fI<clustername>\fP
Create a new, skeleton, configuration file. Note that "create" on its own will 
not create a valid configuration file. Fence agents and nodes will need to be 
added to it before handing it over to ccsd. The new configuration file will
have a version number of 1. Subsequent addnode/delnode/addfence/delfence operations
will increment the version number by 1 each time.
.br
\fIOptions:\fP
.br
.br
\-c <file>         Config file to create. Defaults to /etc/cluster/cluster.conf

.TP 
\fBaddnodeids\fP
Adds node ID numbers to all the nodes in cluster.conf. In RHEL4, node IDs were optional
and assigned by cman when a node joined the cluster. In RHEL5 they must be pre-assigned
in cluster.conf. This command will not change any node IDs that are already set in 
cluster.conf, it will simply add unique node ID numbers to nodes that do not already
have them.


.SH "SEE ALSO"
cluster.conf(5)


rgmanager/ChangeLog
2008-04-16 Fabio M. Di Nitto <fabbione at fabbione.net>
	* src/utils/clustat.c: Fix variable init

2008-03-04 Lon Hohberger <lhh at redhat.com>
	* src/resources/ASEHAagent.sh, Makefile: Add Sybase failover agent
	* src/resources/oracledb.sh, Makefile: Add Oracle 10g failover agent

2008-02-26 Lon Hohberger <lhh at redhat.com>
	* src/resources/ip.sh: Fix netmask handling in ip.sh
	* src/utils/clustat.c: Don't show estranged nodes if they're down
2008-02-01 Lon Hohberger <lhh at redhat.com>
	* src/daemons/rg_state.c, slang_event.c, groups.c, restart_counter.c,
	include/restart_counter.h, resgroup.h,
	src/resources/default_event_script.sl: Allow restart counters to 
	correctly work with central_processing. (#400211 / #431130)

2008-01-25 Lon Hohberger <lhh at redhat.com>
	* src/daemons/rg_thread.c: Fix case that broke 'clusvcadm -e <service>
	-n <node>' case #430220
	* src/daemons/rg_state.c: Don't enable migrate-to-self (#430272)

2008-01-02 Fabio M. Di Nitto <fabbione at ubuntu.com>
	* src/resources/ocf-shellfuncs: Add interpreter.

2007-12-30 Fabio M. Di Nitto <fabbione at ubuntu.com>
	* src/daemons/rg_state.c, include/resgroup.h, src/clulib/alloc.c:
	  Fix building when -DDEBUG is defined.

2007-12-24 Fabio M. Di Nitto <fabbione at ubuntu.com>
	* src/clulib/vft.c: Change ifdef to fix build on parisc.

2007-12-19 Lon Hohberger <lhh at redhat.com>
	* include/res-ocf.h: Fix typo
	* src/daemons/restree.c: Export reference counts to resource agents.
	* src/resources/clusterfs.sh: Use reference counts to ensure we
	don't unmount a GFS file system if it's in use by multiple services
	on the same node.  #254111
	* src/resources/service.sh, default_event_script.sl: Add support for
	"soft" dependencies if central_processing is enabled

2007-12-14 Lon Hohberger <lhh at redhat.com>
	(Misc central_events fixes)
	* Add return code for inability to run because of exclusive 
	tag
	* Only initialize state once
	* Fix master-node failover.
	* Correctly return node after enable if request was forwarded

2007-12-12 Lon Hohberger <lhh at redhat.com>
	* Misc changes; add missing ds.h
	* src/resources/default*.sl: Make clusvcadm -r go to a different
	node.  Don't notice() that we're starting a service if it's
	disabled.
	* src/daemons/groups.c: Make clustat not interfere with startup
	when central processing is enabled
	* src/daemons/slang_event.c: Make clusvcadm correctly report
	service owner after relocate/restart/enable requests.

2007-11-30 Lon Hohberger <lhh at redhat.com>
	* Commit RIND / S-Lang script engine [untested]

[RHEL5 merged ChangeLog Entries]
2007-11-30 Lon Hohberger <lhh at redhat.com>
	* src/resources/clusterfs.sh: Retry mount up to 3 times to avoid
	race condition during another process mounting a GFS volume
	* src/resources/vm.sh, service.sh: Add defaults for values.
	Make vm.sh work with more service attrs (max restarts)
	* src/utils/clustat.c: Make output of clustat terminal-width 
	dependent

2007-11-26 Lon Hohberger <lhh at redhat.com>
	* include/reslist.h: Add restart counters to resource node structure
	(intended for top-level resources, i.e. services, vms...)
	* include/restart_counter.h: Add header file for restart counter
	* src/daemons/Makefile: Fix build to include restart counters
	* src/daemons/restart_counter.c: Implement restart counters #247139
	* src/daemons/fo_domain.c, groups.c, restart_counter.c, resrules.c,
	restree.c, test.c: Glue for restart counters.
	* src/daemons/reslist.c: Glue for restart counters.  Make expand_time
	parser more robust to allow things like '1h30m' as a time value.
	* src/daemons/main.c: Mark quorum disk offline in the correct
	place to avoid extraneous log messages
	* src/daemons/rg_state.c: Allow marking service as stopped if
	stuck in recover state.  Make service which failed to start
	go to stopped state.  Glue for restart counters.
	* src/resources/service.sh, vm.sh: Add parameters for restart
	counters #247139

2007-11-14 Lon Hohberger <lhh at redhat.com>
	* src/utils/clulog.c: Make clulog honor rgmanager log levels
	(#289501)
	* src/clulib/vft.c: Fix #303981 - crash on rgmanager restart in some
	cases
	* man/clusvcadm.8: Remove references to clushutdown from man page;
	resolves #324151
	* src/resources/netfs.sh: Apply patch from Marco Ceci to fix #358161
	* src/resources/vm.sh: Make default migration policy live instead
	of pause for Xen virtual machines.  Also make it configurable instead
	of static.  Resolves #345871

2007-11-13 Lon Hohberger <lhh at redhat.com>
	* src/resources/clusterfs.sh: Add support for self_fence operation
	to clusterfs resource agent
	* src/resources/service.sh: Add default values to service.sh

2007-10-26 Lon Hohberger <lhh at redhat.com>
	* src/daemons/main.c, src/utils/clustat.c, clusvcadm.c:
	Fix #354391

2007-09-25 Lon Hohberger <lhh at redhat.com>
	* src/daemons/restree.c: Apply patch to fix side case re: 229650
	Patch from Simone Gotti. Resolves: #229650
[End RHEL5 merged changes]

2007-08-30 Lon Hohberger <lhh at redhat.com>
	* src/daemons/restree.c, rg_state.c: Fix tree-restart bug
	This is another part of #229650

2007-08-22 Fabio M. Di Nitto <fabbione at ubuntu.com>
	* include/list.h: Fix build with gcc-4.2.

2007-08-15 Lon Hohberger <lhh at redhat.com>
	* src/utils/clustat.c: Remove uninitialized var & fix logic /
	spotted by Ryan McCabe

2007-08-09  Fabio M. Di Nitto <fabbione at ubuntu.com>
	* rgmanager/src/clulib/alloc.c: Undefine DEBUG when building on
	PARISC. __builtin_address needs the same love as IA64.

2007-08-02 Lon Hohberger <lhh at redhat.com>
	* general: More fixes around #248727
	* include/reslist.h, src/daemons/restree.c: Make last-value be
	returned or resources which have been checked recently
	* src/daemons/groups.c: Make VMs use migrate semantics instead of
	relocate semantics when employing failover domain rules
	* src/daemons/nodeevent.c: Fix VMs ending up on wrong nodes when
	simultaneous boot occurs
	* src/daemons/rg_forward.c: Fix erroneous timeout

config/tools/ccs_tool/ccs_tool.c
#include <stdio.h>
#include <stdlib.h>
#include <getopt.h>
#include <string.h>
#include <errno.h>

#include "copyright.cf"
#include "editconf.h"
#ifdef LEGACY_CODE
#include "update.h"
#include "libccscompat.h"
#else
#include "ccs.h"
#endif


/*
 * Old libccs retruned -error (mostly!) but didn't set errno (sigh)
 * New libccs sets errno correctly
 */
static char *errstring(int retcode)
{
#ifdef LEGACY_CODE
	return strerror(retcode);
#else
	return strerror(errno);
#endif
}

static void tool_print_usage(FILE *stream);

int globalverbose=0;

static void test_print_usage(FILE *stream);

static int test_main(int argc, char *argv[], int old_format){
  int desc=0;
  int i=0;
  int error = 0;
  int force = 0, blocking = 0;
  char *str=NULL;
  char *cluster_name = NULL;

  if(argc <= 1){
    test_print_usage(stderr);
    exit(EXIT_FAILURE);
  }

  for(i=1; i < argc; i++){
    if(!strcmp(argv[i], "-h")){
      test_print_usage(stdout);
      exit(EXIT_SUCCESS);
    }
    if(!strcmp(argv[i], "-V")){
      printf("%s %s (built %s %s)\n", argv[0], RELEASE_VERSION, __DATE__, __TIME__);
      printf("%s\n", REDHAT_COPYRIGHT);
      exit(EXIT_SUCCESS);
    }
  }

  if(!strcmp(argv[1], "connect")){
    for(i=2; i < argc; i++){
      if(!strcmp(argv[i], "force")){
	printf("Force is set.\n");
	force = 1;
      } else if(!strcmp(argv[i], "block")){
	printf("Blocking is set.\n");
	blocking = 1;
      } else {
	cluster_name = argv[i];
	printf("Setting cluster name to %s\n", cluster_name);
      }
    }
    if(blocking && !force){
      fprintf(stderr, "Blocking can only be used with \"force\".\n");
      exit(EXIT_FAILURE);
    }
    if(force){
      desc = ccs_force_connect(cluster_name, blocking);
    } else {
      if(cluster_name){
	fprintf(stderr, "A cluster name can only be specified when using 'force'.\n");
	exit(EXIT_FAILURE);
      }
      desc = ccs_connect();
    }
    if(desc < 0){
      fprintf(stderr, "ccs_connect failed: %s\n", errstring(-desc));
      exit(EXIT_FAILURE);
    } else {
      printf("Connect successful.\n");
      printf(" Connection descriptor = %d\n", desc);
#ifndef LEGACY_CODE
      ccs_disconnect(desc);
#endif
    }
  }
  else if(!strcmp(argv[1], "disconnect")){
    if(argc < 3){
      fprintf(stderr, "Wrong number of arguments.\n");
      exit(EXIT_FAILURE);
    }
#ifdef LEGACY_CODE
    desc = atoi(argv[2]);
#else
    desc = ccs_connect();
#endif
    if((error = ccs_disconnect(desc))){
      fprintf(stderr, "ccs_disconnect failed: %s\n", errstring(-error));
      exit(EXIT_FAILURE);
    } else {
      printf("Disconnect successful.\n");
    }
  }
  else if(!strcmp(argv[1], "get")){
    if(argc < 4){
      fprintf(stderr, "Wrong number of arguments.\n");
      exit(EXIT_FAILURE);
    }
#ifdef LEGACY_CODE
    desc = atoi(argv[2]);
#else
    desc = ccs_connect();
#endif
    if((desc < 0) || (error = ccs_get(desc, argv[3], &str))){
      fprintf(stderr, "ccs_get failed: %s\n", errstring(-error));
      exit(EXIT_FAILURE);
    } else {
	    if (old_format) {
		    printf("Get successful.\n");
		    printf(" Value = <%s>\n", str);
	    }
	    else {
		    printf("%s\n", str);
	    }
      if(str)free(str);
#ifndef LEGACY_CODE
      ccs_disconnect(desc);
#endif
    }
  }
  else {
    fprintf(stderr, "Unknown command: %s\n", argv[1]);
    exit(EXIT_FAILURE);
  }

  exit(EXIT_SUCCESS);
}

static void test_print_usage(FILE *stream)
{
	fprintf(stream,
	  "Usage:\n"
	  "\n"
	  "ccs_test [Options] <Command>\n"
	  "\n"
	  "Options:\n"
	  "  -h                        Print usage.\n"
	  "  -V                        Print version information.\n"
	  "\n"
	  "Commands:\n"
	  "  connect <force> <block>   Connect to CCS and return connection descriptor.\n"
	  "  disconnect <desc>         Disconnect from CCS.\n"
	  "  get <desc> <request>      Get a value from CCS.\n"
	  );
}

#ifndef LEGACY_CODE
static int xpath_query(int argc, char **argv)
{
	int handle;
	char *ret;
	int i;

	if (argc < 2) {
		fprintf(stderr,
			"Usage:\n"
			"\n"
			"ccs_tool query [-n] <xpath query>\n"
			"\n"
		        "options\n"
			"  -n     disable full XPath parsing\n");
		return 1;
	}

	/* Tell the library we want full XPath parsing */
	fullxpath = 1;

	if (strcmp(argv[1], "-n") == 0) {
		argv++;
		argc--;

		/* Actually ... no we don't */
		fullxpath = 0;
	}

	handle = ccs_connect();

	/* Process all the queries on the command-line */
	for (i=1; i<argc; i++) {
		if (!ccs_get(handle, argv[1], &ret)) {
			printf("%s\n", ret);
			free(ret);
		}
		else {
			fprintf(stderr, "Query failed: %s\n", strerror(errno));
			ccs_disconnect(handle);
			return -1;
		}
	}
	ccs_disconnect(handle);
	return 0;
}
#endif

static int tool_main(int argc, char *argv[])
{
  optind = 1;

  if (argc < 2 || !strcmp(argv[optind], "-h")) {
      tool_print_usage(stdout);
      exit(EXIT_SUCCESS);
  }
  if (!strcmp(argv[optind], "-V")) {
      printf("%s %s (built %s %s)\n", argv[0], RELEASE_VERSION,
	     __DATE__, __TIME__);
      printf("%s\n", REDHAT_COPYRIGHT);
      exit(EXIT_SUCCESS);
  }

  if(optind < argc){
    if(!strcmp(argv[optind], "-verbose")){
      optind++;
      globalverbose=1;
    }
    if(!strcmp(argv[optind], "help")){
      tool_print_usage(stdout);
      exit(EXIT_SUCCESS);
    }
#ifdef LEGACY_CODE
    /* Update is meaningless now */
    else if(!strcmp(argv[optind], "update")){
      if(optind+1 >= argc){
	fprintf(stderr, "Too few arguments.\n"
		"Try 'ccs_tool help' for help.\n");
	exit(EXIT_FAILURE);
      }
      if(update(argv[optind+1])){
	fprintf(stderr, "\nFailed to update config file.\n");
	exit(EXIT_FAILURE);
      }
      printf("\nUpdate complete.\n");
    }
    /* Do old ccs queries */
    else if(!strcmp(argv[optind], "query")){
	    char *new_argv[argc+2];
	    int i;

	    new_argv[0] = "ccs_test";
	    new_argv[1] = "get";
	    new_argv[2] = "0"; /* Dummy connection ID */
	    for (i=2; i<argc; i++)
		    new_argv[1+i] = argv[i];

	    return test_main(argc+1, new_argv, 0);
    }
#else
    else if(!strcmp(argv[optind], "query")){
	    return xpath_query(argc-1, argv+1);
    }
#endif
    else if(!strcmp(argv[optind], "addnode")){
	    add_node(argc-1, argv+1);
	    exit(EXIT_SUCCESS);
    }
    else if(!strcmp(argv[optind], "delnode")){
	    del_node(argc-1, argv+1);
	    exit(EXIT_SUCCESS);
    }
    else if(!strcmp(argv[optind], "addfence")){
	    add_fence(argc-1, argv+1);
	    exit(EXIT_SUCCESS);
    }
    else if(!strcmp(argv[optind], "delfence")){
	    del_fence(argc-1, argv+1);
	    exit(EXIT_SUCCESS);
    }
    else if(!strcmp(argv[optind], "lsnode")){
	    list_nodes(argc-1, argv+1);
	    exit(EXIT_SUCCESS);
    }
    else if(!strcmp(argv[optind], "lsfence")){
	    list_fences(argc-1, argv+1);
	    exit(EXIT_SUCCESS);
    }
    else if(!strcmp(argv[optind], "create")){
	    create_skeleton(argc-1, argv+1);
	    exit(EXIT_SUCCESS);
    }
    else if(!strcmp(argv[optind], "addnodeids")){
	    add_nodeids(argc-1, argv+1);
	    exit(EXIT_SUCCESS);
    }

    else {
      fprintf(stderr, "Unknown command, %s.\n"
	      "Try 'ccs_tool help' for help.\n", argv[optind]);
      exit(EXIT_FAILURE);
    }
  } else {
    fprintf(stderr, "Too few arguments.\n"
	    "Try 'ccs_tool help' for help.\n");
    exit(EXIT_FAILURE);
  }
  exit(EXIT_SUCCESS);
}

static void tool_print_usage(FILE *stream){
  fprintf(stream,
	  "Usage:\n"
	  "  ccs_tool [options] <command>\n"
	  "\n"
	  "Options:\n"
	  "  -verbose            Make some operations print more details.\n"
	  "  -h                  Print this usage and exit.\n"
	  "  -V                  Print version information and exit.\n"
	  "\n"
	  "Commands:\n"
	  "  help                Print this usage and exit.\n"
#ifdef LEGACY_CODE
	  "  update <xml file>   Tells ccsd to upgrade to new config file version.\n"
	  "  query <ccs query>   Query the cluster configuration.\n"
#else
	  "  query <xpath query> Query the cluster configuration.\n"
#endif
	  "  addnode <node>      Add a node\n"
          "  delnode <node>      Delete a node\n"
          "  lsnode              List nodes\n"
          "  lsfence             List fence devices\n"
	  "  addfence <fencedev> Add a new fence device\n"
	  "  delfence <fencedev> Delete a fence device\n"
	  "  create              Create a skeleton config file\n"
	  "  addnodeids          Assign node ID numbers to all nodes\n"
	  "\n");
}


int main(int argc, char *argv[])
{
	char *name = strdup(argv[0]);

	/*
	 * Don't be anal about the binary name.
	 * We expect either 'ccs_tool' or 'ccs_test',
	 * but interpret anything other than 'ccs_test'
	 * as 'ccs_tool'.
	 * That's not a bug, it's a feature.
	 */

	if (strcmp(basename(name), "ccs_test") == 0)
		return test_main(argc, argv, 1);
	else
		return tool_main(argc, argv);
}
